### ISSUE TITLE
Detect Redis port collisions during health check

### ISSUE THREAD

**Issue by lthibault** [testground/testground#959]
## Description

As discussed in #954, multiple Redis instances will cause cryptic failures in Testground when using the `local:exec` runner.  The issue appears to unfold as follows:

- Health-check is able to connect to a Redis process, so passes.
- Redis container is started by Testground with default port, which collides with the other Redis process.
- Go SDK's `sync` package complains with `failed to create redis client: no viable redis host found`. 

We should make reasonable attempts to detect running Redis instances, and report this in the health check.

## Proposed approach

The simplest approach is to check whether port 6379 is in use (e.g. with `netcat`).  If so, the health-check should fail.

@raulk Any recommendations for:

1. How to detect that a port is in use?  I imagine we want this to be cross-platform.
2. Where in the source code this logic should actually live? (I'm not yet very familiar with the code base).

## What defines this endeavour to be complete?

- [ ] Detect port 6379 in use during health check
- [ ] Health check fails with useful warning

**Comment by hacdias** [comment]
Some notes: I am trying to reproduce this error before trying to fix it. I understand there may be collisions on the port being used by multiple redis instances and that's the cause of the problem.

However, I am running two Redis instances. One on docker:

```shell
$ docker ps
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS                     NAMES
c5395b6f10a1        34df9616b38f        "testground sidecar …"   3 days ago          Up 3 days           0.0.0.0:32769->6060/tcp   testground-sidecar
cbd0f04508ff        influxdb:1.8        "/entrypoint.sh infl…"   3 days ago          Up 3 days           0.0.0.0:8086->8086/tcp    testground-influxdb
74488e24a95a        redis               "docker-entrypoint.s…"   3 days ago          Up 3 days           0.0.0.0:6379->6379/tcp    testground-redis
b13e87533db8        bitnami/grafana     "/run.sh"                3 days ago          Up 3 days           0.0.0.0:3000->3000/tcp    testground-grafana
f2c0a6797bbf        goproxy/goproxy     "/goproxy"               4 days ago          Up 4 days           8081/tcp                  testground-goproxy
```

```
$ docker inspect testground-redis
[
    {
        "Id": "74488e24a95ab12fad318c13835666f2b43811e3ec5aa2c8508d20ef3a65fe5c",
        "Created": "2020-07-17T13:54:43.1501949Z",
        "Path": "docker-entrypoint.sh",
        "Args": [
            "--save",
            "",
            "--appendonly",
            "no",
            "--maxclients",
            "120000",
            "--stop-writes-on-bgsave-error",
            "no"
        ],
        "State": {
            "Status": "running",
            "Running": true,
            "Paused": false,
            "Restarting": false,
            "OOMKilled": false,
            "Dead": false,
            "Pid": 23474,
            "ExitCode": 0,
            "Error": "",
            "StartedAt": "2020-07-17T13:54:43.5882322Z",
            "FinishedAt": "0001-01-01T00:00:00Z"
        },
        "Image": "sha256:2355926154447ec75b25666ff5df14d1ab54f8bb4abf731be2fcb818c7a7f145",
        "ResolvConfPath": "/var/lib/docker/containers/74488e24a95ab12fad318c13835666f2b43811e3ec5aa2c8508d20ef3a65fe5c/resolv.conf",
        "HostnamePath": "/var/lib/docker/containers/74488e24a95ab12fad318c13835666f2b43811e3ec5aa2c8508d20ef3a65fe5c/hostname",
        "HostsPath": "/var/lib/docker/containers/74488e24a95ab12fad318c13835666f2b43811e3ec5aa2c8508d20ef3a65fe5c/hosts",
        "LogPath": "/var/lib/docker/containers/74488e24a95ab12fad318c13835666f2b43811e3ec5aa2c8508d20ef3a65fe5c/74488e24a95ab12fad318c13835666f2b43811e3ec5aa2c8508d20ef3a65fe5c-json.log",
        "Name": "/testground-redis",
        "RestartCount": 0,
        "Driver": "overlay2",
        "Platform": "linux",
        "MountLabel": "",
        "ProcessLabel": "",
        "AppArmorProfile": "",
        "ExecIDs": null,
        "HostConfig": {
            "Binds": null,
            "ContainerIDFile": "",
            "LogConfig": {
                "Type": "json-file",
                "Config": {}
            },
            "NetworkMode": "testground-control",
            "PortBindings": {
                "6379/tcp": [
                    {
                        "HostIp": "",
                        "HostPort": "6379"
                    }
                ]
            },
            "RestartPolicy": {
                "Name": "unless-stopped",
                "MaximumRetryCount": 0
            },
            "AutoRemove": false,
            "VolumeDriver": "",
            "VolumesFrom": null,
            "CapAdd": null,
            "CapDrop": null,
            "Capabilities": null,
            "Dns": null,
            "DnsOptions": null,
            "DnsSearch": null,
            "ExtraHosts": null,
            "GroupAdd": null,
            "IpcMode": "private",
            "Cgroup": "",
            "Links": null,
            "OomScoreAdj": 0,
            "PidMode": "",
            "Privileged": false,
            "PublishAllPorts": false,
            "ReadonlyRootfs": false,
            "SecurityOpt": null,
            "UTSMode": "",
            "UsernsMode": "",
            "ShmSize": 67108864,
            "Sysctls": {
                "net.core.somaxconn": "150000",
                "net.netfilter.nf_conntrack_max": "120000"
            },
            "Runtime": "runc",
            "ConsoleSize": [
                0,
                0
            ],
            "Isolation": "",
            "CpuShares": 0,
            "Memory": 0,
            "NanoCpus": 0,
            "CgroupParent": "",
            "BlkioWeight": 0,
            "BlkioWeightDevice": null,
            "BlkioDeviceReadBps": null,
            "BlkioDeviceWriteBps": null,
            "BlkioDeviceReadIOps": null,
            "BlkioDeviceWriteIOps": null,
            "CpuPeriod": 0,
            "CpuQuota": 0,
            "CpuRealtimePeriod": 0,
            "CpuRealtimeRuntime": 0,
            "CpusetCpus": "",
            "CpusetMems": "",
            "Devices": null,
            "DeviceCgroupRules": null,
            "DeviceRequests": null,
            "KernelMemory": 0,
            "KernelMemoryTCP": 0,
            "MemoryReservation": 0,
            "MemorySwap": 0,
            "MemorySwappiness": null,
            "OomKillDisable": false,
            "PidsLimit": null,
            "Ulimits": [
                {
                    "Name": "nofile",
                    "Hard": 1048576,
                    "Soft": 1048576
                }
            ],
            "CpuCount": 0,
            "CpuPercent": 0,
            "IOMaximumIOps": 0,
            "IOMaximumBandwidth": 0,
            "MaskedPaths": [
                "/proc/asound",
                "/proc/acpi",
                "/proc/kcore",
                "/proc/keys",
                "/proc/latency_stats",
                "/proc/timer_list",
                "/proc/timer_stats",
                "/proc/sched_debug",
                "/proc/scsi",
                "/sys/firmware"
            ],
            "ReadonlyPaths": [
                "/proc/bus",
                "/proc/fs",
                "/proc/irq",
                "/proc/sys",
                "/proc/sysrq-trigger"
            ]
        },
        "GraphDriver": {
            "Data": {
                "LowerDir": "/var/lib/docker/overlay2/376d266b7b4a192e48ecd71892257c4f2daf66d49701883654b695d06a754214-init/diff:/var/lib/docker/overlay2/121e3ea5445b665c00a9d82dbbe303a70e2b662ce4045c588be7bba3e63d7a41/diff:/var/lib/docker/overlay2/9f6e436f65d8e1bce10a9d01f53ed721989e52f588d5a86edf71c9a2cc7e0e9b/diff:/var/lib/docker/overlay2/1618593270f2a4b927273f9b45c515622e611829dd7365729abd898ae649438f/diff:/var/lib/docker/overlay2/9f41859c5ec928cbde6a212f9c83314cec57648610f7015dc0e2683dfacb18a7/diff:/var/lib/docker/overlay2/dc477ea6e0d8685bdf47df889ea09a39c0df4bc57300a4366ff85bc0dfb9458c/diff:/var/lib/docker/overlay2/eef7869d09688adc3190b5b2f68c78433e0a23f54b10d5b924f5c9c4c027cc60/diff",
                "MergedDir": "/var/lib/docker/overlay2/376d266b7b4a192e48ecd71892257c4f2daf66d49701883654b695d06a754214/merged",
                "UpperDir": "/var/lib/docker/overlay2/376d266b7b4a192e48ecd71892257c4f2daf66d49701883654b695d06a754214/diff",
                "WorkDir": "/var/lib/docker/overlay2/376d266b7b4a192e48ecd71892257c4f2daf66d49701883654b695d06a754214/work"
            },
            "Name": "overlay2"
        },
        "Mounts": [
            {
                "Type": "volume",
                "Name": "2305d52f99b830b0b1f16a17537e27422a420117ff8266f6d02aa872b9a642c8",
                "Source": "/var/lib/docker/volumes/2305d52f99b830b0b1f16a17537e27422a420117ff8266f6d02aa872b9a642c8/_data",
                "Destination": "/data",
                "Driver": "local",
                "Mode": "",
                "RW": true,
                "Propagation": ""
            }
        ],
        "Config": {
            "Hostname": "74488e24a95a",
            "Domainname": "",
            "User": "",
            "AttachStdin": false,
            "AttachStdout": false,
            "AttachStderr": false,
            "ExposedPorts": {
                "6379/tcp": {}
            },
            "Tty": false,
            "OpenStdin": false,
            "StdinOnce": false,
            "Env": [
                "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
                "GOSU_VERSION=1.12",
                "REDIS_VERSION=6.0.5",
                "REDIS_DOWNLOAD_URL=http://download.redis.io/releases/redis-6.0.5.tar.gz",
                "REDIS_DOWNLOAD_SHA=42cf86a114d2a451b898fcda96acd4d01062a7dbaaad2801d9164a36f898f596"
            ],
            "Cmd": [
                "--save",
                "",
                "--appendonly",
                "no",
                "--maxclients",
                "120000",
                "--stop-writes-on-bgsave-error",
                "no"
            ],
            "Image": "library/redis",
            "Volumes": {
                "/data": {}
            },
            "WorkingDir": "/data",
            "Entrypoint": [
                "docker-entrypoint.sh"
            ],
            "OnBuild": null,
            "Labels": {}
        },
        "NetworkSettings": {
            "Bridge": "",
            "SandboxID": "08376a342a7585de2be10815d47147c158a8801e7a39956e985e8ce77b6967f0",
            "HairpinMode": false,
            "LinkLocalIPv6Address": "",
            "LinkLocalIPv6PrefixLen": 0,
            "Ports": {
                "6379/tcp": [
                    {
                        "HostIp": "0.0.0.0",
                        "HostPort": "6379"
                    }
                ]
            },
            "SandboxKey": "/var/run/docker/netns/08376a342a75",
            "SecondaryIPAddresses": null,
            "SecondaryIPv6Addresses": null,
            "EndpointID": "",
            "Gateway": "",
            "GlobalIPv6Address": "",
            "GlobalIPv6PrefixLen": 0,
            "IPAddress": "",
            "IPPrefixLen": 0,
            "IPv6Gateway": "",
            "MacAddress": "",
            "Networks": {
                "testground-control": {
                    "IPAMConfig": null,
                    "Links": null,
                    "Aliases": [
                        "74488e24a95a"
                    ],
                    "NetworkID": "00999e3f4605a0c0fb85b07b3b7c3c7c9145fe2ad3d833d976d08681bc0ae843",
                    "EndpointID": "add9ede0a2f80133f949c96a0e8e85ed4c561cf33bd3c73ea64d12f911f30c2c",
                    "Gateway": "192.18.0.1",
                    "IPAddress": "192.18.0.3",
                    "IPPrefixLen": 16,
                    "IPv6Gateway": "",
                    "GlobalIPv6Address": "",
                    "GlobalIPv6PrefixLen": 0,
                    "MacAddress": "02:42:c0:12:00:03",
                    "DriverOpts": null
                }
            }
        }
    }
]

```

And one other through homebrew:

```
$ brew services list
Name       Status  User         Plist
caddy      stopped
emacs      stopped
mysql      stopped
parity     stopped
php        stopped
postgresql started henriquedias /Users/henriquedias/Library/LaunchAgents/homebrew.mxcl.postgresql.plist
redis      started henriquedias /Users/henriquedias/Library/LaunchAgents/homebrew.mxcl.redis.plist
unbound    stopped
```

```
$ cat /usr/local/etc/redis.conf | grep ^port
port 6379
```

Both seem to be running on the same port: 6379.

```
$ telnet localhost 6379
Trying ::1...
Connected to localhost.
Escape character is '^]'.
ping
+PONG
quit
+OK
Connection closed by foreign host.
```

However, running the network/ping-pong plan using the `local:exec` runner works without major complaining. I checked and the ping-pong plan actually uses the `sync` package as expected.

```
$ testground run single \
         --plan=network \
         --testcase=ping-pong \
         --builder=exec:go \
         --runner=local:exec \
         --instances=2
Jul 21 09:19:15.484480	[34mINFO[0m	created a synthetic composition file for this job; all instances will run under singleton group "single"
Jul 21 09:19:15.484536	[34mINFO[0m	using home directory: /Users/henriquedias/testground
Jul 21 09:19:15.484569	[34mINFO[0m	no .env.toml found at /Users/henriquedias/testground/.env.toml; running with defaults
Jul 21 09:19:15.484575	[34mINFO[0m	testground client initialized	{"addr": "localhost:8042"}
Jul 21 09:19:15.484970	[34mINFO[0m	using home directory: /Users/henriquedias/testground
Jul 21 09:19:15.485003	[34mINFO[0m	no .env.toml found at /Users/henriquedias/testground/.env.toml; running with defaults
Jul 21 09:19:15.485007	[34mINFO[0m	testground client initialized	{"addr": "localhost:8042"}
Jul 21 09:19:15.485121	[34mINFO[0m	test plan source at: /Users/henriquedias/testground/plans/network
[1;36m
>>> Server output:
[0m
Jul 21 09:19:15.491660	[34mINFO[0m	performing build for groups	{"req_id": "217e9a3c", "plan": "network", "groups": ["single"], "builder": "exec:go"}
Jul 21 09:19:16.874324	[34mINFO[0m	build succeeded	{"req_id": "217e9a3c", "plan": "network", "groups": ["single"], "builder": "exec:go", "artifact": "/Users/henriquedias/testground/data/work/exec-go--network-65373b2828af"}
[1;92m
>>> Result:
[0m
Jul 21 09:19:16.874935	[34mINFO[0m	generated build artifact	{"group": "single", "artifact": "/Users/henriquedias/testground/data/work/exec-go--network-65373b2828af"}
[1;36m
>>> Server output:
[0m
Jul 21 09:19:16.875216	[34mINFO[0m	performing healthcheck on runner	{"req_id": "62bdb8a2"}
Jul 21 09:19:16.901887	[34mINFO[0m	[1;32mhealthcheck: ok[0m	{"req_id": "62bdb8a2"}
Jul 21 09:19:16.902420	[34mINFO[0m	starting test case instance	{"req_id": "62bdb8a2", "plan": "network", "group": "single", "number": 0, "total": 1}
Jul 21 09:19:16.905332	[34mINFO[0m	starting test case instance	{"req_id": "62bdb8a2", "plan": "network", "group": "single", "number": 1, "total": 2}
Jul 21 09:19:17.339922	[34mINFO[0m	0.4377s [30;47m   MESSAGE[0m [31m<< single[000] >>[0m registering default http handler at: http://[::]:6060/ (pprof: http://[::]:6060/debug/pprof/)	{"req_id": "62bdb8a2"}
Jul 21 09:19:17.340187	[34mINFO[0m	0.4378s [30;106m     START[0m [31m<< single[000] >>[0m {"plan":"network","case":"ping-pong","run":"2ba8d9eda8f1","outputs_path":"/Users/henriquedias/testground/data/outputs/local_exec/network/2ba8d9eda8f1/single/0","instances":2,"group":"single","group_instances":2,"network":"127.1.0.0/16","start_time":"0001-01-01T00:00:00Z"}	{"req_id": "62bdb8a2"}
Jul 21 09:19:17.340259	[34mINFO[0m	0.4380s [30;47m   MESSAGE[0m [31m<< single[000] >>[0m before sync.MustBoundClient	{"req_id": "62bdb8a2"}
Jul 21 09:19:17.340361	[34mINFO[0m	0.4377s [30;47m   MESSAGE[0m [32m<< single[001] >>[0m registering default http handler at: http://[::]:57453/ (pprof: http://[::]:57453/debug/pprof/)	{"req_id": "62bdb8a2"}
Jul 21 09:19:17.340434	[34mINFO[0m	0.4378s [30;106m     START[0m [32m<< single[001] >>[0m {"plan":"network","case":"ping-pong","run":"2ba8d9eda8f1","outputs_path":"/Users/henriquedias/testground/data/outputs/local_exec/network/2ba8d9eda8f1/single/1","instances":2,"group":"single","group_instances":2,"network":"127.1.0.0/16","start_time":"0001-01-01T00:00:00Z"}	{"req_id": "62bdb8a2"}
Jul 21 09:19:17.340471	[34mINFO[0m	0.4381s [30;47m   MESSAGE[0m [32m<< single[001] >>[0m before sync.MustBoundClient	{"req_id": "62bdb8a2"}
Jul 21 09:19:17.345408	[34mINFO[0m	0.4432s [37;42m        OK[0m [31m<< single[000] >>[0m 	{"req_id": "62bdb8a2"}
Jul 21 09:19:17.345437	[34mINFO[0m	0.4432s [37;42m        OK[0m [32m<< single[001] >>[0m 	{"req_id": "62bdb8a2"}
Jul 21 09:19:17.389314	[34mINFO[0m	0.4871s [30;47m   MESSAGE[0m [32m<< single[001] >>[0m io closed	{"req_id": "62bdb8a2"}
Jul 21 09:19:17.389535	[34mINFO[0m	0.4871s [30;47m   MESSAGE[0m [31m<< single[000] >>[0m io closed	{"req_id": "62bdb8a2"}
Jul 21 09:19:17.402661	[34mINFO[0m	0.5005s [30;47m   MESSAGE[0m [31m<< single[000] >>[0m influxdb: uploaded 2 points	{"req_id": "62bdb8a2"}
Jul 21 09:19:17.403043	[34mINFO[0m	0.5010s [30;47m   MESSAGE[0m [32m<< single[001] >>[0m influxdb: uploaded 2 points	{"req_id": "62bdb8a2"}
Jul 21 09:19:17.424731	[34mINFO[0m	run finished successfully	{"req_id": "62bdb8a2", "run_id": "2ba8d9eda8f1", "plan": "network", "case": "ping-pong", "runner": "local:exec", "instances": 2}
[1;92m
>>> Result:
[0m
Jul 21 09:19:17.425048	[34mINFO[0m	finished run with ID: 2ba8d9eda8f1
```

So I'm not being able to reproduce the problem. @raulk @nonsense any pointers here?

**Comment by raulk** [comment]
@hacdias in this case the user had a pre-existing instance of Redis running locally that had already bound to port 6379, so the `local:exec` runner failed to start the Redis docker container on the same port (as we use host port mapping to expose the container to local processes over port 6379).

There's nothing technically wrong here, but it's an opportunity to improve our UX.

The idea here is to add a healthcheck rule to `local:exec` that tries to bind to port 6379 on localhost, by setting up a [TCP listener](https://golang.org/pkg/net/#ListenTCP). On error, we fail the healthcheck and report a user-friendly message like "local port 6379 is already occupied; please stop any local Redis instances first." This healthcheck has no associated fixer.

**Comment by hacdias** [comment]
Got it! Makes sense. I will make sure to do that then!

**Comment by nonsense** [comment]
I think we have introduced a regression with this one.

```
ubuntu@ip-172-31-3-94:~$ sudo lsof -i -P -n | grep LISTEN
systemd-r  1038 systemd-resolve   13u  IPv4   3717      0t0  TCP 127.0.0.53:53 (LISTEN)
sshd       1561            root    3u  IPv4   3977      0t0  TCP *:22 (LISTEN)
sshd       1561            root    4u  IPv6   3979      0t0  TCP *:22 (LISTEN)
testgroun 15532          ubuntu    6u  IPv4 104504      0t0  TCP 127.0.0.1:8042 (LISTEN)
docker-pr 27431            root    4u  IPv6  65267      0t0  TCP *:3000 (LISTEN)
docker-pr 27677            root    4u  IPv6  67706      0t0  TCP *:6379 (LISTEN)
docker-pr 27942            root    4u  IPv6  67778      0t0  TCP *:8086 (LISTEN)
docker-pr 28095            root    4u  IPv6  67835      0t0  TCP *:32768 (LISTEN)
ubuntu@ip-172-31-3-94:~$ docker ps
CONTAINER ID        IMAGE                       COMMAND                  CREATED             STATUS              PORTS                     NAMES
0add81259062        iptestground/sidecar:edge   "testground sidecar …"   59 minutes ago      Up 59 minutes       0.0.0.0:32768->6060/tcp   testground-sidecar
37ae048b9bc0        influxdb:1.8                "/entrypoint.sh infl…"   About an hour ago   Up 59 minutes       0.0.0.0:8086->8086/tcp    testground-influxdb
97a8d9c0cc28        redis                       "docker-entrypoint.s…"   About an hour ago   Up About an hour    0.0.0.0:6379->6379/tcp    testground-redis
1bd5c28153b5        bitnami/grafana             "/run.sh"                About an hour ago   Up About an hour    0.0.0.0:3000->3000/tcp    testground-grafana
4fff5500e642        goproxy/goproxy             "/goproxy"               About an hour ago   Up About an hour    8081/tcp                  testground-goproxy
```

testground-redis container is running, but when trying to start a test plan, we get:

```
>>> Server output:

Sep 10 10:53:47.322080  INFO    performing healthcheck on runner        {"req_id": "fba7df09"}
Sep 10 10:53:47.328745  WARN    engine run error: healthcheck fixes failed; aborting:
Checks:
- redis-port: failed; local port 6379 is already occupied; please stop any local Redis instances first.
- local-outputs-dir: ok; directory exists.
- control-network: ok; network exists.
- local-grafana: ok; container state: running
- local-redis: ok; container state: running
- local-influxdb: ok; container state: running
Fixes:
- redis-port: failed; requires manual fixing.
- local-outputs-dir: unnecessary;
- control-network: unnecessary;
- local-grafana: unnecessary;
- local-redis: unnecessary;
- local-influxdb: unnecessary;
        {"req_id": "fba7df09"}

>>> Error:

engine run error: healthcheck fixes failed; aborting:
Checks:
- redis-port: failed; local port 6379 is already occupied; please stop any local Redis instances first.
- local-outputs-dir: ok; directory exists.
- control-network: ok; network exists.
- local-grafana: ok; container state: running
- local-redis: ok; container state: running
- local-influxdb: ok; container state: running
Fixes:
- redis-port: failed; requires manual fixing.
- local-outputs-dir: unnecessary;
- control-network: unnecessary;
- local-grafana: unnecessary;
- local-redis: unnecessary;
- local-influxdb: unnecessary;
```

System is: `Linux 5.3.0-1034-aws #36-Ubuntu SMP Tue Aug 18 08:58:43 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux`

**Comment by hacdias** [comment]
I'm closing this as it seems to have been fixed by #1210 which fixed #1202 which seems to be exactly this seem issue.
